core/new-gui/src/app/workspace/service/workflow-graph/model/shared-model-change-handler.ts

```diff
@@ -301,38 +303,70 @@ export class SharedModelChangeHandler {
               operatorID: operatorID,
               newDisplayName: newName.toJSON(),
             });
-          } else if (event.path[event.path.length - 1] === "inputPorts") {
-            const addedInputPort = event.delta[1]?.insert ? event.delta[1]?.insert : event.delta[0]?.insert;
-            if (addedInputPort) {
-              this.onOperatorPortAdded(
-                operatorID,
-                true,
-                (addedInputPort as Y.Map<any>[])[0].toJSON() as PortDescription
-              );
-            }
-            if (event.delta[0]?.delete || event.delta[1]?.delete) {
-              this.onOperatorPortRemoved(operatorID, true);
-            }
-          } else if (event.path[event.path.length - 1] === "outputPorts") {
-            const addedOutputPort = event.delta[1]?.insert ? event.delta[1]?.insert : event.delta[0]?.insert;
-            if (addedOutputPort) {
-              this.onOperatorPortAdded(
-                operatorID,
-                false,
-                (addedOutputPort as Y.Map<any>[])[0].toJSON() as PortDescription
-              );
-            }
-            if (event.delta[0]?.delete || event.delta[1]?.delete) {
-              this.onOperatorPortRemoved(operatorID, false);
-            }
           } else if (event.path.includes("operatorProperties")) {
             this.onOperatorPropertyChanged(operatorID, event.transaction.local);
+          } else if (event.path.includes("inputPorts")) {
+            this.handlePortEvent(event, operatorID, true);
+          } else if (event.path.includes("outputPorts")) {
+            this.handlePortEvent(event, operatorID, false);
+          } else {
+            throw new Error(`undefined operation on shared type: .${event}`);
           }
         }
       });
     });
   }
 
+  /**
+   * Handles the additon, deletion and deeper changes to operator ports.
+   * @param event
+   * @param operatorID
+   * @param isInput Since input and output ports are separate properties, need to access them differently.
+   * @private
+   */
+  private handlePortEvent(event: Y.YEvent<Y.Map<any>>, operatorID: string, isInput: boolean) {
+    if (event.path.length === 2) {
+      // Port added or deleted
+      const addedPort = event.delta[1]?.insert ? event.delta[1]?.insert : event.delta[0]?.insert;
+      if (addedPort) {
+        this.onPortAdded(operatorID, isInput, (addedPort as Y.Map<any>[])[0].toJSON() as PortDescription);
+      }
+      if (event.delta[0]?.delete || event.delta[1]?.delete) {
+        this.onPortRemoved(operatorID, isInput);
+      }
```

Comment:
can one port be "added" and "deleted" in the same event? if not maybe change to "else if", or handle all other conditions explicitly?

Reply:
No, changed.